From 4132f8717b5083edecd2353951a386714588a36b Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Mon, 21 Nov 2011 22:16:25 +0000
Subject: [PATCH 05/22] OS-742 zoneadmd crash leaves zones unable to shutdown

---
 usr/src/cmd/zoneadmd/mcap.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/usr/src/cmd/zoneadmd/mcap.c b/usr/src/cmd/zoneadmd/mcap.c
index db6ce68..3330da3 100644
--- a/usr/src/cmd/zoneadmd/mcap.c
+++ b/usr/src/cmd/zoneadmd/mcap.c
@@ -585,7 +585,6 @@ again:
 			free(buf);
 			return (0);
 		}
-		goto again;
 	}
 
 	if (n > got) {
-- 
2.8.0

